### ISSUE TITLE
cannot detect "referenced before assignment" static rule

### ISSUE THREAD

**Issue by MYTHOFLEGEND** [microsoft/pyright#5174]
**Describe the bug**
cannot detect referenced before assignment error sometimes

**Screenshots or Code**
![image](https://github.com/microsoft/pyright/assets/48516484/da9e821e-cae2-4186-bd3c-7f6c23532530)

![image](https://github.com/microsoft/pyright/assets/48516484/781d4f98-f500-418d-8105-36261444b912)

in the first image, vscode detected that error, but in the second one, vscode cannot detected this with the same code stucture

![image](https://github.com/microsoft/pyright/assets/48516484/c1682851-fe4f-4ce9-bd99-5c36c7b0da27)

and this line should at least show a warning, because if a == 1, video_path is not defined

**Comment by erictraut** [comment]
Please provide a self-contained, minimal code sample (in text form) that demonstrates the problem. Reproing and debugging issues from partial screen shots isn't possible.

The code in your screen shots also references symbols that are not built in, so it's not possible to determine how you've defined them.

**Comment by MYTHOFLEGEND** [comment]
![image](https://github.com/microsoft/pyright/assets/48516484/49b26cb7-d79e-4de4-a9b8-ff1fc8cf8bde)


```python
def func():
    flag = 0
    
    if flag == 0:
        video_path = ""
        
    elif flag == 1:
        long_video_path = video_path
        
    a = video_path
    b = long_video_path
```

**Comment by MYTHOFLEGEND** [comment]
![image](https://github.com/microsoft/pyright/assets/48516484/a8612d34-7509-4798-8152-8b60d477283b)

```python
from random import random, randint

def add_numbers(number_1: int, number_2: int) -> int:
    return number_1 + number_2

if __name__ == "__main__":
    random_value: float = random()

    if random_value > 0.5:
        result: int = add_numbers(number_1=randint(1, 49), number_2=randint(50, 100))
    
    else:
        print("Addition skipped.")

    new_result: int = result + 1
    print(f"New result is: {new_result}")
```

**Comment by erictraut** [comment]
I suspect this is because you have `python.analysis.typeCheckingMode` set to "off". If you set this to "basic", pylance will report these errors. Alternatively, you can enable type checking in a single file by adding a `# pyright: basic` or `# pyright: strict` comment at the top of the file. Does that fix your issue?

![image](https://github.com/microsoft/pyright/assets/7040122/978b7470-7da7-4a82-b37c-4e4885aa6d7d)

**Comment by MYTHOFLEGEND** [comment]
it worked, thank you very much!
